src/main/java/org/prebid/server/analytics/reporter/greenbids/GreenbidsAnalyticsReporter.java

```diff
@@ -0,0 +1,355 @@
+        }
+
+        if (bidResponse == null || auctionContext == null) {
+            return Future.failedFuture("Bid response or auction context cannot be null");
+        }
+
+        final String greenbidsId = UUID.randomUUID().toString();
+        final String billingId = UUID.randomUUID().toString();
+
+        final Boolean isSampled = isSampled(greenbidsAnalyticsProperties.getGreenbidsSampling(), greenbidsId);
+
+        if (!isSampled) {
+            return Future.succeededFuture();
+        }
+
+        final CommonMessage commonMessage;
+        try {
+            commonMessage = createBidMessage(
+                    auctionContext,
+                    bidResponse,
+                    greenbidsId,
+                    billingId);
+        } catch (IllegalArgumentException e) {
+            return Future.failedFuture(e);
+        }
+
+        try {
+            final String commonMessageJson = jacksonMapper.encodeToString(commonMessage);
+
+            final MultiMap headers = MultiMap.caseInsensitiveMultiMap()
+                    .add(HttpUtil.ACCEPT_HEADER, HttpHeaderValues.APPLICATION_JSON)
+                    .add(HttpUtil.CONTENT_TYPE_HEADER, HttpHeaderValues.APPLICATION_JSON);
+
+            final Future<HttpClientResponse> responseFuture = httpClient.post(
+                    greenbidsAnalyticsProperties.getAnalyticsServer(),
+                    headers,
+                    commonMessageJson,
+                    greenbidsAnalyticsProperties.getTimeoutMs());
+
+            return responseFuture.compose(response -> {
+                final int responseStatusCode = response.getStatusCode();
+                if (responseStatusCode == 202 || responseStatusCode == 200) {
+                    return Future.succeededFuture();
+                } else {
+                    return Future.failedFuture(
+                            "Unexpected response status: " + response.getStatusCode());
+                }
+            });
+
+        } catch (EncodeException e) {
+            return Future.failedFuture("Failed to encode as JSON: " + e.getMessage());
+        }
+
+    }
+
+    private Boolean isSampled(double samplingRate, String greenbidsId) {
+        if (samplingRate < 0 || samplingRate > 1) {
+            logger.warn("Warning: Sampling rate must be between 0 and 1");
+        }
+
+        final double exploratorySamplingRate = samplingRate
+                * greenbidsAnalyticsProperties.getExploratorySamplingSplit();
+        final double throttledSamplingRate = samplingRate
+                * (1.0 - greenbidsAnalyticsProperties.getExploratorySamplingSplit());
+
+        final long hashInt = Integer.parseInt(
+                greenbidsId.substring(greenbidsId.length() - 4), 16);
+        final boolean isPrimarySampled = hashInt < exploratorySamplingRate * RANGE_16_BIT_INTEGER_DIVISION_BASIS;
+
+        return isPrimarySampled
+                || hashInt >= (1 - throttledSamplingRate) * RANGE_16_BIT_INTEGER_DIVISION_BASIS;
```

Comment:
please extract the condition `hashInt >= (1 - throttledSamplingRate) * RANGE_16_BIT_INTEGER_DIVISION_BASIS` as a separate variable (like the `isPrimarySample`) with a clear name